cpp/src/io/parquet/page_enc.cu

```diff
@@ -1337,11 +1345,196 @@ static __device__ void byte_reverse128(__int128_t v, void* dst)
+    if (not strings::detail::is_valid_begin_utf8_char(str[idx])) { return false; }
+    // subsequent bytes in the characater should be continuation bytes
+    auto const width = strings::detail::bytes_in_utf8_byte(str[idx++]);
+    for (size_type i = 1; i < width && idx < str.size_bytes(); i++, idx++) {
+      if (not strings::detail::is_utf8_continuation_char(str[idx])) { return false; }
+    }
+  }
+
+  return true;
+}
+
+/**
+ * @brief Increment part of a UTF-8 character.
+ *
+ * Attempt to increment the byte pointed to by ptr, which is assumed to be part of a valid UTF-8
+ * character. Returns true if successful, false if the increment caused an overflow, in which case
+ * the data at ptr will be set to the lowest valid UTF-8 byte (start or continuation).
+ * Will halt execution if passed invalid UTF-8.
+ */
+__device__ bool increment_utf8_at(uint8_t* ptr)
+{
+  uint8_t elem = *ptr;
+  // elem is one of (no 5 or 6 byte chars allowed):
+  //  0b0vvvvvvv 1 byte
+  //  0b10vvvvvv continuation
+  //  0b110vvvvv 2 byte
+  //  0b1110vvvv 3 byte
+  //  0b11110vvv 4 byte
+
+  // TODO(ets): starting at 4 byte and working down.  Should probably start low and work higher.
+  uint8_t mask  = 0xF8;
+  uint8_t valid = 0xF0;
+
+  while (mask != 0) {
+    if ((elem & mask) == valid) {
+      elem++;
+      if ((elem & mask) != mask) {  // no overflow
+        *ptr = elem;
+        return true;
+      }
+      *ptr = valid;
+      return false;
+    }
+    mask <<= 1;
+    valid <<= 1;
+  }
+
+  // should not reach here since we test for valid UTF-8 higher up the call chain
+  CUDF_UNREACHABLE("Trying to increment non-utf8");
+}
+
+/**
+ * @brief Truncate a UTF-8 string to at most truncate_length bytes.
+ *
+ * If is_min is false, then the final character (or characters if there is overflow) will be
+ * incremented so that the resultant string will still be a valid maximum. scratch is only used when
+ * is_min is false, and must be at least truncate_length bytes in size. If the string cannot be
+ * truncated, leave it untouched and return the original length.
+ *
+ * @return Pair object containing a pointer to the truncate string and its length.
+ */
+__device__ std::pair<const void*, uint32_t> truncate_utf8(device_span<uint8_t const> str,
+                                                          bool is_min,
+                                                          void* scratch,
+                                                          size_type truncate_length)
+{
+  // we know at this point that truncate_length < size_bytes, so
+  // there is a character at [len]. work backwards until we find
+  // the start of a unicode character.
+  auto len  = truncate_length;
+  auto dptr = &str.data()[len];
+  while (not strings::detail::is_begin_utf8_char(*dptr) && len > 0) {
+    dptr--;
+    len--;
+  }
+  if (len != 0) {
+    if (is_min) { return {str.data(), len}; }
+    memcpy(scratch, str.data(), len);
```

Comment:
This is unfortunate, but I don't see a way around it as we may end up mutating all the values to round it up for a max. Consider the binary `0xEFFFFF...FFFF` being truncated, it would end up 0xFF0000...00`.

Reply:
Or worse, `0xFFFFFF...FF` can't be truncated at all, and should be left alone, but if you just mutated every byte...
So scratch it must be.